File: admin/src/lib/resources/payroll.ts
import { api, downloadFile } from "@/lib/api";
import { PayrollSummary, PayrollDetail, PayrollDefectComparison } from "@/lib/types";

export interface PayrollSlip {
  _id: string;
  worker: { _id: string; code: string; name: string };
  periodFrom: string;
  periodTo: string;
  totalQuantity: number;
  totalAmount: number;
  reportCount: number;
  issuedAt: string;
}

export const payrollApi = {
  summary: (from: string, to: string) => api.get<PayrollSummary>(`/payroll/summary?from=${from}&to=${to}`),
  detail: (workerId: string, from: string, to: string) =>
    api.get<PayrollDetail>(`/payroll?worker=${workerId}&from=${from}&to=${to}`),
  defectComparison: (workerId: string, from: string, to: string) =>
    api.get<PayrollDefectComparison>(`/payroll/defect-comparison?worker=${workerId}&from=${from}&to=${to}`),
  export: (worker: string, from: string, to: string) =>
    api.post<PayrollSlip>(`/payroll/export`, { worker, from, to }),
  listSlips: (params?: { worker?: string; from?: string; to?: string }) => {
    const qs = new URLSearchParams();
    if (params?.worker) qs.set("worker", params.worker);
    if (params?.from) qs.set("from", params.from);
    if (params?.to) qs.set("to", params.to);
    const suffix = qs.toString() ? `?${qs.toString()}` : "";
    return api.get<PayrollSlip[]>(`/payroll/slips${suffix}`);
  },
  downloadSlipFile: (id: string, format: "pdf" | "xlsx", fileName: string) =>
    downloadFile(`/payroll/slips/${id}/export?format=${format}`, fileName),
};

File: app/lib/models/payroll.dart
import 'worker.dart';
import 'production_report.dart';

class PayrollRow {
  final Worker? worker;
  final double totalQuantity;
  final double totalAmount;
  final int reportCount;

  PayrollRow({
    required this.worker,
    required this.totalQuantity,
    required this.totalAmount,
    required this.reportCount,
  });

  factory PayrollRow.fromJson(Map<String, dynamic> json) => PayrollRow(
        worker: json['worker'] != null ? Worker.fromJson(json['worker'] as Map<String, dynamic>) : null,
        totalQuantity: (json['totalQuantity'] as num).toDouble(),
        totalAmount: (json['totalAmount'] as num).toDouble(),
        reportCount: json['reportCount'] as int? ?? 0,
      );
}

class PayrollSummary {
  final String from;
  final String to;
  final List<PayrollRow> rows;

  PayrollSummary({required this.from, required this.to, required this.rows});

  factory PayrollSummary.fromJson(Map<String, dynamic> json) => PayrollSummary(
        from: json['from'] as String,
        to: json['to'] as String,
        rows: (json['rows'] as List<dynamic>).map((e) => PayrollRow.fromJson(e as Map<String, dynamic>)).toList(),
      );
}

class PayrollSlip {
  final String id;
  final DateTime periodFrom;
  final DateTime periodTo;
  final double totalQuantity;
  final double totalAmount;
  final int reportCount;
  final DateTime issuedAt;

  PayrollSlip({
    required this.id,
    required this.periodFrom,
    required this.periodTo,
    required this.totalQuantity,
    required this.totalAmount,
    required this.reportCount,
    required this.issuedAt,
  });

  factory PayrollSlip.fromJson(Map<String, dynamic> json) => PayrollSlip(
        id: json['_id'] as String,
        periodFrom: DateTime.parse(json['periodFrom'] as String),
        periodTo: DateTime.parse(json['periodTo'] as String),
        totalQuantity: (json['totalQuantity'] as num).toDouble(),
        totalAmount: (json['totalAmount'] as num).toDouble(),
        reportCount: json['reportCount'] as int? ?? 0,
        issuedAt: DateTime.parse(json['issuedAt'] as String),
      );
}

class PayrollDetail {
  final String from;
  final String to;
  final double totalQuantity;
  final double totalAmount;
  final List<ProductionReport> reports;

  PayrollDetail({
    required this.from,
    required this.to,
    required this.totalQuantity,
    required this.totalAmount,
    required this.reports,
  });

  factory PayrollDetail.fromJson(Map<String, dynamic> json) => PayrollDetail(
        from: json['from'] as String,
        to: json['to'] as String,
        totalQuantity: (json['totalQuantity'] as num).toDouble(),
        totalAmount: (json['totalAmount'] as num).toDouble(),
        reports: (json['reports'] as List<dynamic>)
            .map((e) => ProductionReport.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}
